### ISSUE TITLE
Configuration problem with python

### ISSUE THREAD

**Issue by ngupta26** [FluidityProject/fluidity#271]
Hello
I am facing an issue in build process related to python. I am not an expert in installation and I am new to opensource, so please go easy on my mistakes. Thank You
About the issue,
----
I had downloaded and extracted the source code. 
./configure : step is facing an error as shown below

configure: Checking for Python automagic.
configure: No Python automagic.
checking for python... /home/nikhil/anaconda3/bin/python
checking for a version of Python >= '2.1.0'... yes
checking for the distutils Python package... yes
checking for Python include path... -I/home/nikhil/anaconda3/include/python3.6m
checking for Python library path... -L/home/nikhil/anaconda3/lib/python3.6/config -lpython3.6
checking for Python site-packages path... /home/nikhil/anaconda3/lib/python3.6/site-packages
checking python extra libraries...  -lpthread -ldl  -lutil -lrt
checking python extra linking flags... -Xlinker -export-dynamic
checking consistency of all components of python development environment... no
configure: error: in `/home/nikhil/soft/fluidity-4.1.11':
configure: error: 
  Could not link test program to Python. Maybe the main Python library has been
  installed in some non-standard library path. If so, pass it to configure,
  via the LDFLAGS environment variable.
  Example: ./configure LDFLAGS="-L/usr/non-standard-path/python/lib"
---
But what I gathered from this was that it was able to capture the path for python and the main issue was in
"checking consistency of all components of python development environment... no".

So I looked into the config.log file and what I gathered was an issue in line 590:
/usr/bin/ld: cannot find -lpython3.6

Please find the config.log file in the attachment.

Can some one please help. I am afraid that this issue is because I have python through Anaconda.

Thank You
[config.log](https://github.com/FluidityProject/fluidity/files/4794003/config.log)

**Comment by gnikit** [comment]
Hi @ngupta26 the easiest way to get started with Fluidity is to use the ppa, which installs all the necessary libraries (Python, PETSc, Zoltan, Metis, SCOTCH, etc.) for configuring and compiling, see [here](https://launchpad.net/~fluidity-core/+archive/ubuntu/ppa).
```sh
sudo add-apt-repository ppa:fluidity-core/ppa
sudo apt-get update
sudo apt install fluidity-dev-python3
```

This should hopefully fix your problem. If you have done that already and it didn't work or you want to manually install all the dependencies we can have a more detailed look.

**Comment by stephankramer** [comment]
Hi @ngupta it looks like you're not using master fluidity, as the configure has changed. Are you using IC-FERST? Also I see that you are on an Ubuntu system, is there a reason you can't use the system python?
The actual error is telling you that it's looking for a file `libpython3.6.so` or possibly `libpython3.6m.so` and you need to tell it where it is using:
```
./configure LDFLAGS="-L<path where you've found libpython>"
```

**Comment by ngupta26** [comment]
Hi @gnikit , Thanks for helping out.
Initially, I had started with ppa only. I followed the steps as mentioned in the manual

```shell
sudo apt-add-repository -y ppa:fluidity-core/ppa
sudo apt-get update
sudo apt-get -y install fluidity
```

Everything went smoothly. I checked "fluidity -V" to find that it has python support 
````shell
PETSc support			yes
Python support			yes
Numpy support			yes
````
but I was unable to run the examples as given in the manual
````shell
cd /tmp
tar -zxvf /usr/share/doc/fluidity/examples.tar.gz
cd exaples/top_hat/
````
on make prepocess, i faced the following issue
````shell
$ make preprocess
*** ERROR ***
Fluidity python support, required by this example, was not found on your
system. Please either (a) install the Fluidity binary package if you are on
Ubuntu, or (b) get hold of a Fluidity source tree and set your PYTHONPATH to
refer to this, as described in the Fluidity manual.
envcheck.mk:45: recipe for target 'envcheck' failed
make: *** [envcheck] Error 1
````
Thus, I gave up then and decided to remove fluidiity and build it myself. 

Regarding your suggestion
> 
> ```shell
> sudo add-apt-repository ppa:fluidity-core/ppa
> sudo apt-get update
> sudo apt install fluidity-dev-python3
> ```
> 
> This should hopefully fix your problem. If you have done that already and it didn't work or you want to manually install all the dependencies we can have a more detailed look.

After your reply, i tried it again with the statement "sudo apt install fluidity-dev-python3"
````shell
sudo add-apt-repository ppa:fluidity-core/ppa
sudo apt-get update
sudo apt install fluidity-dev-python3
sudo apt-get -y install fluidity
````
But even now the problem persists. On doing make preprocess in the top_hat directory as above, I faced the same issue
````shell
*** ERROR ***
Fluidity python support, required by this example, was not found on your
system. Please either (a) install the Fluidity binary package if you are on
Ubuntu, or (b) get hold of a Fluidity source tree and set your PYTHONPATH to
refer to this, as described in the Fluidity manual.
envcheck.mk:45: recipe for target 'envcheck' failed
make: *** [envcheck] Error 1
````
Should I continue via apt-get or should I try to build it manually only? 

Thank You
Nikhil

**Comment by gnikit** [comment]
If memory serves me right installing fluidity (`sudo apt install fluidity`) from the PPA installs only the binary of fluidity, which is the application. The `fluidity-dev-python3`/ `fluidity-dev` on the other hand install the dependencies for you, in order to configure and then build (`make`) Fluidity yourself.

> `sudo apt-get -y install fluidity`

This installs the fluidity binary, so if you want to build fluidity yourself you have to skip this step.

Having said that you can follow the instructions below to build Fluidity from sources.
Having installed `fluidity-dev-python3` or `fluidity-dev` and using the system python not anaconda you should then do something like this in a new terminal window to download, configure and build fluidity.

```bash
git clone https://github.com/FluidityProject/fluidity.git
cd fluidity
./configure --enable-2d-adaptivity
make all -j
```

Now, the working version of Fluidity should be in `./fluidity/bin/fluidity`, add that to your PATH.

**Comment by Patol75** [comment]
You may also need to edit your PYTHONPATH environment variable after having built from source for the Python script included in Fluidity to be available. On my system, I have:
`/home/thomas/Softwares/fluidity/tools:/home/thomas/Softwares/fluidity/python`.

**Comment by ngupta26** [comment]
> Hi @ngupta it looks like you're not using master fluidity, as the configure has changed. Are you using IC-FERST? Also I see that you are on an Ubuntu system, is there a reason you can't use the system python?
> The actual error is telling you that it's looking for a file `libpython3.6.so` or possibly `libpython3.6m.so` and you need to tell it where it is using:
> 
> ```
> ./configure LDFLAGS="-L<path where you've found libpython>"
> ```

Hi @stephankramer , Thank you for the early reply and help.
Sorry for the late reply as I was trying to install fluidity as suggested and then ask all my question.

Yes, I wasn't using the master fluidity and hence was facing the issue. 
Also, unfortunately, there were some issue few years back with my system python as far as I can recall, hence I had resorted to anaconda since then. It is fairly possible that it would have been corrected by now, but I am not sure.

But after using master fluidity, this current python issue has been resolved.

Thank you very much

**Comment by ngupta26** [comment]
@gnikit @stephankramer  @Patol75 
Thank You everyone for your help. After using master fluidity as suggested by Stephan, this issue was resolved. I am definitely further ahead in installing fluidity as I was yesterday. Due to the uncertainties in my system, I have decided to install fluidity in a Singularity container. 

I am facing few issues in it, I will post them as a new issue if I am unable to solve them myself.
Thank You very much for the support and help.

Cheers
Nikhil
